### PR TITLE
Add expires_after ppx

### PR DESCRIPTION
**Pull request by psteckler**
RFC 0012 for RPC versioning mentions a ppx to enforce expiring old versions.

This is an implementation of that ppx. It generates a compile-time error if you compile past the given expiration date (or the date doesn't comply with the format).

### CODE REVIEW

File: src/lib/ppx_coda/expires_after.ml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  in
+  let today = Date.today ~zone:Time.Zone.utc in
+  if Date.( >= ) today date then Location.raise_errorf ~loc "Code is expired"
+  else
+    [%expr
+      let _ = () in
+      ()]
+
+let ext =
+  Extension.declare name Extension.Context.expression
```
[the remark this answers is not in the document]
Comment (psteckler) [reply src/lib/ppx_coda/expires_after.ml]: Ah, that's what I was looking for, thanks.

File: src/lib/ppx_coda/expires_after.ml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+let expand ~loc ~path:_ s =
+  let date =
+    try Date.of_string_iso8601_basic s ~pos:0 with _ ->
+      Location.raise_errorf ~loc "Not a valid date; must be in form YYYYMMDD"
+  in
+  let today = Date.today ~zone:Time.Zone.utc in
+  if Date.( >= ) today date then Location.raise_errorf ~loc "Code is expired"
+  else
+    [%expr
+      let _ = () in
```
[the remark this answers is not in the document]
Comment (psteckler) [reply src/lib/ppx_coda/expires_after.ml]: Obviated by your other suggestion.

**Comment by cmr** [comment]
Can we have standalone tests/examples for the PPX?

**Comment by psteckler** [comment]
> Can we have standalone tests/examples for the PPX?

I think those would have to be scripts, which I can add.

**Comment by psteckler** [comment]
Made ppx work as structure item, per @mrmr1993. Added tests, per @cmr.

**Comment by psteckler** [comment]
Note that the ppx causes a failure a compile-time, so you're forced to remove the offending code.

**Comment by bkase** [comment]
Can you add the tests to CI?

**Comment by bkase** [comment]
It can be part of `build_dev` I think

**Comment by psteckler** [comment]
> It can be part of `build_dev` I think

added to `test_all` script

File: scripts/test_all.sh [hunk 3]
```diff
@@ -17,6 +17,11 @@ run_unit_tests() {
   run_dune runtest --verbose -j${MYPROCS}
 }
 
+run_ppx_tests() {
+  date
+  make ppx_tests
```
Comment (mrmr1993) [inline scripts/test_all.sh]: @psteckler shouldn't this be `make test-ppx` to match the makefile?

**Pull request merged**